File: SaveTheCat.py
#initial περιέχει τον αρχικό πίνακα όπως είναι
#waters, cats (list) περιέχει τα νερά/γατες και τα εμπόδια
#time_water, time_cat (list) περιέχει τους χρόνους στους οποίους φτάνει το νερό σε κάθε θέση
#water_queue, cat_queue (queue) είναι η ουρά με τα νερά/γατα
#m (integer) είναι οι στήλες του αρχικού πίνακα
#n (integer) είναι οι γραμμές τους αρχικού πίνακα
#best (list) 2 θέσεων, στην 1η είναι η καλύτρη θέση και στη 2η ο καλύτερος χρόνος
#path (list) θα με βοηθησει να βρω την διαδρομή

def process(f):
    C = list(map(str, f.read().split()))
    n = len(C)
    m = len(C[0])
    initial = list()
    waters = list()
    cats = list()
    time_water = list()
    time_cat = list()
    water_queue = queue.Queue()
    cat_queue = queue.Queue()
    best = [None]*2
    path = [None]*(n*m)
    cat_is_dead = [None]*1
    cat_is_dead[0]=0
    cat_position = 0
    for i in range(0,n):
        line = C[i]
        for j in range(0,m):
            initial.append(line[j])
    for x in range(0,len(initial)):
        waters.append(initial[x])
        cats.append(initial[x])
        time_cat.append(0)
        if initial[x] == "W":
            time_water.append(0)
            water_queue.put(x)
        elif initial[x] == "A":
            cat_queue.put(x)
            cat_position = x
            time_water.append(n*m+1)
        else:
            time_water.append(n*m+1)

    water_flood(water_queue, time_water, waters, cat_is_dead, m, n, cat_position)
    cat_flood(cat_is_dead, best, cat_queue, time_cat, cats, time_water, path, m, n, cat_position)
    result(cat_is_dead, best, cat_position,path, m, n)


import queue

def water_flood(water_queue, time_water, waters, cat_is_dead, m, n, cat_position):
    while water_queue.empty() == False:
        curr = water_queue.get()
        counter = time_water[curr]
        if (curr-m) >= 0:
            if (waters[curr-m] != "X") and (waters[curr-m] != "W"):
                water_queue.put(curr-m)
                time_water[curr-m] = counter+1
                waters[curr-m] = "W"
        if (curr+m) <= n*m-1:
            if (waters[curr+m] != "X") and (waters[curr+m] != "W"):
                water_queue.put(curr+m)
                time_water[curr+m] = counter+1
                waters[curr+m] = "W"
        if (curr%m) != 0:
            if (waters[curr-1] != "X") and (waters[curr-1] != "W"):
                water_queue.put(curr-1)
                time_water[curr-1] = counter+1
                waters[curr-1] = "W"
        if (curr%m) != m-1:
            if (waters[curr+1] != "X") and (waters[curr+1] != "W"):
                water_queue.put(curr+1)
                time_water[curr+1] = counter+1
                waters[curr+1] = "W"
    if waters[cat_position] == "W":
        cat_is_dead[0] = 1

#δεν πρεπει να ξεχασω να βαλώ στο best το αντιστοιχο coordinates.posission
#και επειδή δεν μπορώ να το κάνω εδώ θα το κάνω όταν γίνεται το διαβασμα
#δεν ξεχναω πρεπει να το περασω σαν input

#στον πινακα path πρεπει να βαλω στην αρχικη θεση της γατας μια αρνητικη τιμη

#επίσης να μην ξεχασω να αρχικοποιήσω την πρώτη θέση της γάτας
#κραταω αυτη την τιμη και στο best και στο cat_position

def cat_flood(cat_is_dead, best, cat_queue, time_cat, cats, time_water, path, m, n, cat_position):
    if cat_is_dead[0] == 0:
        best[1] = n*m
    else:
        best[1] = 0
    best[0] = cat_position
    path[cat_position] = -m-1
    while cat_queue.empty() == False:
        curr = cat_queue.get()
        counter = time_cat[curr]
        if curr+m <= n*m-1:
            if (cats[curr+m] != "X") and (cats[curr+m] != "A") and (counter+1 < time_water[curr+m]):
                path[curr+m] = curr
                cats[curr+m] = "A"
                cat_queue.put(curr+m)
                time_cat[curr+m] = counter+1
                if best[1] < time_water[curr+m]-1:
                    best[1] = time_water[curr+m]-1
                    best[0] = curr+m
                elif best[1] == time_water[curr+m]-1:
                    if (curr+m) < best[0]:
                        best[0] = curr+m
        if curr%m != 0:
            if (cats[curr-1] != "X") and (cats[curr-1] != "A") and (counter+1 < time_water[curr-1]):
                path[curr-1] = curr
                cats[curr-1] = "A"
                cat_queue.put(curr-1)
                time_cat[curr-1] = counter+1
                if best[1] < time_water[curr-1]-1:
                    best[1] = time_water[curr-1]-1
                    best[0] = curr-1
                elif best[1] == time_water[curr-1] - 1:
                    if (curr-1) < best[0]:
                        best[0] = curr-1
        if curr%m != m-1:
            if (cats[curr+1] != "X") and (cats[curr+1] != "A") and (counter+1 < time_water[curr+1]):
                path[curr+1] = curr
                cats[curr+1] = "A"
                cat_queue.put(curr+1)
                time_cat[curr+1] = counter+1
                if best[1] < time_water[curr+1]-1:
                    best[1] = time_water[curr+1]-1
                    best[0] = curr+1
                elif best[1] == time_water[curr+1]-1:
                    if (curr+1) < best[0]:
                        best[0] = curr+1
        if curr-m >= 0:
            if (cats[curr-m] != "X") and (cats[curr-m] != "A") and (counter+1 < time_water[curr-m]):
                path[curr-m] = curr
                cats[curr-m] = "A"
                cat_queue.put(curr-m)
                time_cat[curr-m] = counter+1
                if best[1] < time_water[curr-m]-1:
                    best[1] = time_water[curr-m]-1
                    best[0] = curr-m
                elif best[1] == time_water[curr-m]-1:
                    if (curr-m) < best[0]:
                        best[0] = curr-m

def result(cat_is_dead, best, cat_position, path, m, n):
    output = list()
    length = 0
    if cat_is_dead[0] == 1:
        print(best[1])
    else: print("infinity")
    first = best[0]
    if cat_position == first:
        print("stay")
    while True:
        if(path[first] == first-m):
            output.append("D")
            length = length+1
        elif (path[first] == first+m):
            output.append("U")
            length = length+1
        elif (path[first] == first-1):
            output.append("R")
            length = length+1
        elif (path[first] == first+1):
            output.append("L")
            length = length+1
        else: break
        first = path[first]
    if length > 0:
        for i in range(length-1, 0,-1):
            print(output[i], end="")
        print(output[0])

if __name__ == "__main__":
    import sys
    if len(sys.argv) > 1:
        with open(sys.argv[1], "rt") as f:
            process(f)
    else:
        process(sys.stdin)

File: SaveTheCat.cpp
#include <iostream>
#include <queue>
#include <stack>
#include <fstream>
#include <vector>

using namespace std;

struct best{
  int posission;
  int time;
};

void water(queue <int> &nero,vector <int> &wtime,vector <char> &waters, int m, int n, bool &cat_is_dead){
  int counter=0, curr;
  while(!nero.empty()){
    curr = nero.front();
    if ((curr-m) >= 0){
      if((waters[curr-m] != 'X') && (waters[curr-m] != 'W')){
        if (waters[curr-m]=='A') cat_is_dead = true;
        nero.push(curr-m);
        wtime[curr-m] = counter+1;
        waters[curr-m] = 'W';
      }
    }
    if((curr+m) <= n*m-1){
      if((waters[curr+m] != 'X') && (waters[curr+m] != 'W')){
        if (waters[curr+m]=='A') cat_is_dead = true;
        nero.push(curr+m);
        wtime[curr+m] = counter+1;
        waters[curr+m] = 'W';
      }
    }
    if((curr)%m != 0){
      if((waters[curr-1] != 'X') && (waters[curr-1] != 'W')){
        if (waters[curr-1]=='A') cat_is_dead = true;
        nero.push(curr-1);
        wtime[curr-1] = counter+1;
        waters[curr-1] = 'W';
      }
    }
    if((curr)%m != m-1){
      if((waters[curr+1] != 'X') && (waters[curr+1] != 'W')){
        if (waters[curr+1]=='A') cat_is_dead = true;
        nero.push(curr+1);
        wtime[curr+1] = counter+1;
        waters[curr+1] = 'W';
      }
    }
    nero.pop();
    if(!nero.empty())
    counter = wtime[nero.front()];
  }
}

void cat(queue <int> &gata,vector <int> &ctime,vector <int> &wtime,vector <char> &cats, int m, int &n,vector <int> &ex, best &coordinates, bool cat_is_dead){
  int counter=0, curr;
  coordinates.posission = gata.front();
  if (!cat_is_dead) coordinates.time = n*m;
  else coordinates.time = 0;
  ex[gata.front()] = -3;
  while(!gata.empty()){
    curr = gata.front();
    if((curr+m) <= n*m-1){
      if((cats[curr+m] != 'X') && (cats[curr+m] != 'A') && ((counter+1) < wtime[curr+m])){
        ex[curr+m] = curr;
        gata.push(curr+m);
        ctime[curr+m] = counter+1;
        cats[curr+m] = 'A';
        if(coordinates.time < wtime[curr+m]-1){
          coordinates.time = wtime[curr+m]-1;
          coordinates.posission = curr+m;
        }
        else if(coordinates.time == wtime[curr+m]-1){
          if(curr+m < coordinates.posission) coordinates.posission = curr+m;
        }
      }
    }
    if((curr)%m != 0){
      if((cats[curr-1] != 'X') && (cats[curr-1] != 'A') && ((counter+1) < wtime[curr-1])){
        ex[curr-1] = curr;
        gata.push(curr-1);
        ctime[curr-1] = counter+1;
        cats[curr-1] = 'A';
        if(coordinates.time < wtime[curr-1]-1){
          coordinates.time = wtime[curr-1]-1;
          coordinates.posission = curr-1;
        }
        else if(coordinates.time == wtime[curr-1]-1){
          if(curr-1 < coordinates.posission) coordinates.posission = curr-1;
        }
      }
    }
    if((curr)%m != m-1){
      if((cats[curr+1] != 'X') && (cats[curr+1] != 'A') && ((counter+1) < wtime[curr+1])){
        ex[curr+1] = curr;
        gata.push(curr+1);
        ctime[curr+1] = counter+1;
        cats[curr+1] = 'A';
        if(coordinates.time < wtime[curr+1]-1){
          coordinates.time = wtime[curr+1]-1;
          coordinates.posission = curr+1;
        }
        else if(coordinates.time == wtime[curr+1]-1){
          if(curr+1 < coordinates.posission) coordinates.posission = curr+1;
        }
      }
    }
    if ((curr-m) >= 0){
      if((cats[curr-m] != 'X') && (cats[curr-m] != 'A') && ((counter+1) < wtime[curr-m])){
        ex[curr-m] = curr;
        gata.push(curr-m);
        ctime[curr-m] = counter+1;
        cats[curr-m] = 'A';
        if(coordinates.time < wtime[curr-m]-1){
          coordinates.time = wtime[curr-m]-1;
          coordinates.posission = curr-m;
        }
        else if(coordinates.time == wtime[curr-m]-1){
          if(curr-m < coordinates.posission) coordinates.posission = curr-m;
        }
      }
    }
    gata.pop();
    if(!gata.empty())
    counter = ctime[gata.front()];
  }
}

void result(best coordinates, vector <int> ex, int m, bool cat_is_dead, int cat_posission){
  stack <char> output;
  if (cat_is_dead == true) cout << coordinates.time << endl;
  else cout << "infinity\n";
  int first = coordinates.posission;
  if(first == cat_posission) cout << "stay" ;
  while(1){
    if(ex[first] == first-m) output.push('D');
    else if(ex[first] == first+m) output.push('U');
    else if(ex[first] == first-1) output.push('R');
    else if(ex[first] == first+1) output.push('L');
    else break;
    first = ex[first];
  }
  while(!output.empty()){
    cout << output.top();
    output.pop();
  }
  cout << endl;
}

int main(int num, char **text)
{
  ifstream file;
  file.open(text[1]);
  int n=0, m=0, counter=0, cat_posission;
  bool dead = false;
  vector <char> waters, cats;
  queue <int> nero, gata;
  best coordinates;
  char w;
  do{
    file.get(w);
    waters.push_back(w);
    cats.push_back(w);
    counter++;
    if(w == '\n') {
      if(m == 0) m = counter-1; //γιατι ειναι και το \n που εχει πειραξει το counter
      n++;
      waters.pop_back();
      cats.pop_back();
    }
  }while(!file.eof());
  n = n-1; //γιατι το eof βαζει μια παραπανω γραμμη

  vector <int> wtime, ctime, ex;
  ex.resize(n*m);

  for(int i=0; i<n*m; i++){
    if(waters[i] == 'W'){
      wtime.push_back(0);
      nero.push(i);
    }
    else if(waters[i] == 'A'){
     gata.push(i);
    wtime.push_back(n*m+1);
    }
    else wtime.push_back(n*m+1);
    ctime.push_back(0);
  }
  cat_posission = gata.front();
  water(nero, wtime, waters, m, n, dead);
  cat(gata, ctime, wtime, cats, m, n, ex, coordinates, dead);
  result(coordinates, ex, m, dead, cat_posission);
  file.close();
  return 0;
}
